src/relay/pass/pointer_analysis.cc

```diff
@@ -0,0 +1,168 @@
+  Map<AbstractLocation, Expr> origin;
+  Map<Expr, Set<AbstractLocation>> contain;
+  Map<AbstractLocation, Set<Expr>> store;
+  bool progress = true;
+  Analyze(Map<Expr, AbstractLocation> spawn,
+          Map<AbstractLocation, Expr> origin) : spawn(spawn), origin(origin) {
+    for (const auto& p : origin) {
+      store.Set(p.first, Set<Expr>());
+    }
+  }
+
+  void analyze(const Expr& e) {
+    progress = false;
+    VisitExpr(e);
+  }
+
+  void VisitExpr(const Expr& e) final {
+    if (contain.count(e) == 0) {
+      contain.Set(e, Set<AbstractLocation>());
+    }
+    ExprFunctor::VisitExpr(e);
+  }
+
+  void VisitExpr_(const ConstantNode* op) final { }
+  void VisitExpr_(const OpNode* op) final { }
+  void VisitExpr_(const ConstructorNode* op) final { }
+
+  void FlowTo(AbstractLocation from, const Expr& to) {
+    Set<AbstractLocation> locs = contain.at(to);
+    if (locs.count(from) == 0) {
+      progress = true;
+      locs.Insert(from);
+      contain.Set(to, locs);
+    }
+  }
+
+  void FlowTo(const Expr& from, AbstractLocation to) {
+    Set<Expr> exprs = store.at(to);
+    if (exprs.count(from) == 0) {
+      progress = true;
+      exprs.Insert(from);
+      store.Set(to, exprs);
+    }
+  }
+
+  void FlowTo(const Expr& from, const Expr& to) {
+    for (auto loc : contain.at(from)) {
+      FlowTo(loc, to);
+    }
+  }
+
+  void VisitExpr_(const CallNode* op) final {
+    VisitExpr(op->op);
+    for (Expr e : op->args) {
+      VisitExpr(e);
+    }
+    Expr expr = GetRef<Expr>(op);
+    CHECK(op->op.as<OpNode>()) << "Only support call to operator right now.";
+    FlowTo(op->op, expr);
+    for (Expr e : op->args) {
+      FlowTo(e, expr);
+    }
+  }
+
+  void VisitExpr_(const RefCreateNode* op) final {
+    VisitExpr(op->value);
+    Expr expr = GetRef<Expr>(op);
+    FlowTo(spawn.at(expr), expr);
+    FlowTo(op->value, spawn.at(expr));
+  }
+
+  void VisitExpr_(const RefReadNode* op) final {
+    VisitExpr(op->ref);
+    Expr self = GetRef<Expr>(op);
+    CHECK_GT(contain.count(op->ref), 0);
+    std::unordered_set<Expr, NodeHash, NodeEqual> exprs;
+    for (const auto& abs_loc : contain.at(op->ref)) {
+      if (abs_loc == external_ref) {
+        FlowTo(external_ref, self);
+      } else {
+        CHECK_GT(store.count(abs_loc), 0);
+        for (const auto& expr : store.at(abs_loc)) {
+          exprs.insert(expr);
+        }
+      }
+    }
+    for (const auto& expr : exprs) {
+      FlowTo(expr, self);
+    }
+  }
+
+  void VisitExpr_(const RefWriteNode* op) final {
+    VisitExpr(op->ref);
+    VisitExpr(op->value);
+    CHECK_GT(contain.count(op->ref), 0);
+    for (const auto& loc : contain.at(op->ref)) {
+      FlowTo(op->value, loc);
+    }
+  }
+};
+
+PointerAnalysisResult PointerAnalysis(const Expr& e) {
+  AllocAbstractLocation aal;
+  aal(e);
+  Map<Expr, AbstractLocation> spawn = aal.spawn;
+  Map<AbstractLocation, Expr> origin = aal.origin;
+  Analyze ana(spawn, origin);
+  while (ana.progress) {
```

Comment:
I think you should probably put a comment or clarification as to the purpose of the `progress` flag. Maybe intersperse more comments about how your implementation corresponds to the description of the algorithm